Disable cut/remove grass under vehicles

irwiss opened the pull request:
#### Summary
None

#### Purpose of change
Fixes #54317

#### Describe the solution

Disable cut/remove grass under vehicles

EDIT: reverted the vehicle check from can_construct and used empty_check_lite in json instead, I kept some of the tidying up;

can_construct had bitwise &= which [seems to block short circuiting even in trivial cases](https://godbolt.org/z/P8d8bosqj) and queried map::furn/map::ter twice because it shadowed the names of the previous calls - changed it use normal boolean logic instead of bitwise &

Also make explain_fail_map in load_construction static, and select the correct key using the string instead of constructing new map per load_construction call

#### Describe alternatives you've considered

#### Testing

Replicate scenario in #54317 - grass tile under vehicle (or items/monsters/any of the empty_check_lite conditions) shouldn't be selectable for construction.

Walk to a window with a wall near it, select deconstruct, select the wall, should add message "You cannot deconstruct this!"
Make crafting spot, select tile with the wall should add message "You cannot build there"

#### Additional context

andrei8l commented:
There are 2 slightly misnamed `pre_special`s that are supposed to handle this, but some definitions don't use them - `check_empty` (should be `check_empty_flat`) and `check_empty_lite` (should be `check_empty`).

IMO it'd be better to spread those special checks to construction definitions or even make them default, rather than hard-coding a check.

irwiss commented:
Not sure what construction work can be viable under a vehicle but I'll take a fresh look tomorrow

PatrikLundell commented:
Building a roof over an appliance (or vehicle) seems reasonable to me, as long as the floor remains unchanged (that's what I could come up with for now, anyway).

Qrox commented:
I remember there has been a hack to place a vehicle across a door and board up the door to make a walk-in freezer. It would be nice if we can make this an actual feature by allowing constructing special walls on a vehicle tile (think of it as cementing the whole tile into place).

PatrikLundell commented:
You can do it in the reverse order currently by placing an appliance on a wall tile, as with wall wiring. However, placing the appliance first and constructing the wall around it makes a lot of sense for specific applications.

irwiss commented:
I suppose it makes sense to do the json check instead, I reverted the veh_at check and kept/cherry picked some of the c++ changes i had; the bitwise to boolean logic change and made explain_fail_map static instead of constructed for every construction.

Added the pre_special entry for all cut and remove grass constructions.

I also checked if check_empty_lite can be defaulted to as andrei8l suggested but it doesn't look like it, at least not easily since some constructions run the validation checks via flags, others require furniture/terrain.
Making it the new default and overriding with the current would add 318 lines to the json, when there's around 600ish constructions total so probably not worth it either.
